Fixes 242: Copy local_context_data as-is in Record._add_cache()

markkuleinio opened the pull request:
In #242 the problem seems to be that `local_context_data` contents are not properly copied to `Record._init_cache`, so `update()` and `save()` operations fail to recognize the changed contents of `local_context_data`.

This PR adds a check that for `local_context_data` it uses `copy.deepcopy()` to copy the dict contents in the init cache (and all other fields are handled as earlier). `local_context_data` only contains user-entered data so no specific attribute checks should be performed for it.

With this change the test code in https://github.com/digitalocean/pynetbox/issues/242#issuecomment-654988849 works and saves the `local_context_data` change.

@zachmoody feel free to evaluate if some other way to resolve the issue is more correct. Btw, there is an unused `Record._compare()` method but I didn't touch it yet. I can remove it as well if needed. (Update: oops `_compare()` seems to be used in tests I see)

zachmoody commented:
> Btw, there is an unused Record._compare() method but I didn't touch it yet. I can remove it as well if needed. (Update: oops _compare() seems to be used in tests I see)

Yeah, does seem to be unused. Removing it probably wouldn't be an issue.

I like the approach. Might be a good idea to add a test for this in `test_response` for the `_diff` method. Just so we don't get any surprising behavior though.

markkuleinio commented:
I added a change for `local_context_data` field in the existing `test_diff()`, and also changed all the tests that used `_compare()` instead of `_diff()`.

markkuleinio commented:
I tried to change the test to catch the situation prior to `_add_cache()` change and noticed that `Record._parse_values()` did not copy the `local_context_data` at all, so I changed `_parse_values()` as well. I'm still slightly unsure if I'm doing the correct things so let's see 😃

zachmoody commented:
That's strange it didn't even add the attribute as a record object. Makes sense you'd have to update `_parse_values()` as well though. This looks good.

Sorry the tests are mess. I'm (slowly) trying to get to the point where we're mostly unit-testing everything and the test_blah files only do the generic integration tests. They're a bit overloaded right now as you found with the `_compare()` tests. Would've been totally fine to just remove those instead, but not going to complain about a net increase in testing. :grin:

markkuleinio commented:
Thanks, I missed one thing though: the # from the issue number in the title 😂

zachmoody commented:
haha, no worries, closed it out.
